packages/jss/src/plugins/keyframesRule.js

```diff
@@ -0,0 +1,131 @@
+/* @flow */
+import warning from 'warning'
+import RuleList from '../RuleList'
+import type StyleSheet from '../StyleSheet'
+import type {
+  CSSKeyframesRule,
+  JssStyle,
+  Rule,
+  RuleOptions,
+  ToCssOptions,
+  ContainerRule,
+  KeyframesMap
+} from '../types'
+
+const defaultToStringOptions = {
+  indent: 1,
+  children: true
+}
+
+const nameRegExp = /@keyframes\s+([\w-]+)/
+
+/**
+ * Rule for @keyframes
+ */
+export class KeyframesRule implements ContainerRule {
+  type = 'keyframes'
+
+  at: string = '@keyframes'
+
+  key: string
+
+  name: string
+
+  id: string
+
+  rules: RuleList
+
+  options: RuleOptions
+
+  isProcessed: boolean = false
+
+  renderable: ?CSSKeyframesRule
+
+  constructor(key: string, frames: Object, options: RuleOptions) {
+    const nameMatch = key.match(nameRegExp)
+    if (nameMatch && nameMatch[1]) {
+      this.name = nameMatch[1]
+    } else {
+      this.name = 'noname'
+      warning(false, '[JSS] Bad keyframes name %s', key)
+    }
+    this.key = `${this.type}-${this.name}`
+    this.id = this.name
+    this.options = options
+    this.rules = new RuleList({...options, parent: this})
+
+    for (const name in frames) {
+      this.rules.add(name, frames[name], {
+        ...this.options,
+        parent: this,
+        selector: name
+      })
+    }
+
+    this.rules.process()
+  }
+
+  /**
+   * Generates a CSS string.
+   */
+  toString(options?: ToCssOptions = defaultToStringOptions): string {
+    if (options.children === false) {
+      return `${this.at} ${this.id} {}`
+    }
+    let children = this.rules.toString(options)
+    if (children) children = `\n${children}\n`
+    return `${this.at} ${this.id} {${children}}`
+  }
+}
+
+const keyRegExp = /@keyframes\s+/
+
+const refRegExp = /\$([\w-]+)/
+
+/**
+ * Replace the reference for a animation name.
+ */
+const replaceRef = (style: JssStyle, prop: string, keyframes: KeyframesMap) => {
+  const value = style[prop]
+
+  if (Array.isArray(value)) {
```

Comment:
I am not sure we should support arrays here at all, if we would only support animationName, but if we do we def need to test, I already see this has  a bug

Reply:
As far as I understand animation-name position in animation property is fixed: https://www.w3.org/TR/css-animations-1/#animation

Please try in a codesandbox to be sure, but if so we don't need a loop.